cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaLogProcessorPipeline.java

```diff
@@ -235,21 +239,44 @@ protected String getServiceName() {
    */
   private void initializeOffsets() throws InterruptedException {
     // Setup initial offsets
-    Set<Integer> partitions = config.getPartitions();
-    while (offsets.size() != partitions.size() && !stopped) {
-      for (int partition : partitions) {
+    Set<Integer> partitions = new HashSet<>();
+    partitions.addAll(config.getPartitions());
+    Map<Integer, Integer> partitionRetryCount = new HashMap<>();
+    while (!partitions.isEmpty() && !stopped) {
+      Iterator<Integer> iterator = partitions.iterator();
+      while (iterator.hasNext()) {
+        int partition = iterator.next();
         Checkpoint checkpoint = checkpoints.get(partition);
-        if (checkpoint != null) {
-          // TODO: (CDAP-7684) Deal with wrong offset issue
-          offsets.put(partition, checkpoint.getNextOffset());
-          continue;
-        }
-
         try {
+          if (checkpoint != null) {
+            long offset = offsetResolver.getMatchingOffset(checkpoint, partition);
+            if (offset > 0) {
+              offsets.put(partition, offset);
+              continue;
+            }
+          }
           // If no checkpoint, fetch from the beginning.
           offsets.put(partition, getLastOffset(partition, kafka.api.OffsetRequest.EarliestTime()));
+          // Remove the partition successfully stored in offsets to avoid unnecessary retry for this partition
+          partitions.remove(partition);
         } catch (Exception e) {
-          OUTAGE_LOG.info("Failed to get Kafka earliest offset in {}:{} for pipeline {}. Will be retried",
+          Integer previousRetryCount = partitionRetryCount.get(partition);
+          if (previousRetryCount == null) {
+            partitionRetryCount.put(partition, 1);
+          } else {
+            // (previousRetryCount + 1) is the current number of retries
+            if (previousRetryCount + 1 == KAFKA_INIT_OFFSET_RETRY_TIMES) {
```

Comment:
is 5 second retry enough ? what if few partitions initialize failed, we wouldn't process from them while processing messages from other partitions which successfully initialized. users might skip seeing that error and later figure out some messages are missing ?